daemon/daemon.go

```diff
@@ -97,91 +97,12 @@ type Command struct {
 	GET  ResponseFunc
 	PUT  ResponseFunc
 	POST ResponseFunc
-	// can guest GET?
-	GuestOK bool
-	// can non-admin GET?
-	UserOK bool
-	// is this path accessible on the snapd-snap socket?
-	SnapOK bool
-	// this path is only accessible to root
-	RootOnly bool
-
-	// can polkit grant access? set to polkit action ID if so
-	PolkitOK string
 
-	d *Daemon
-}
-
-// canAccess checks the following properties:
-//
-// - if the user is `root` everything is allowed
-// - if a user is logged in (via `snap login`) and the command doesn't have RootOnly, everything is allowed
-// - POST/PUT all require `root`, or just `snap login` if not RootOnly
-//
-// Otherwise for GET requests the following parameters are honored:
-// - GuestOK: anyone can access GET
-// - UserOK: any uid on the local system can access GET
-// - RootOnly: only root can access this
-// - SnapOK: a snap can access this via `snapctl`
-func (c *Command) canAccess(r *http.Request, user *auth.UserState) accessResult {
-	if c.RootOnly && (c.UserOK || c.GuestOK || c.SnapOK) {
-		// programming error
-		logger.Panicf("Command can't have RootOnly together with any *OK flag")
-	}
-
-	if user != nil && !c.RootOnly {
-		// Authenticated users do anything not requiring explicit root.
-		return accessOK
-	}
-
-	ucred, err := ucrednetGet(r.RemoteAddr)
-	if err != nil && err != errNoID {
-		logger.Noticef("unexpected error when attempting to get UID: %s", err)
-		return accessForbidden
-	}
-	// isUser means we have a UID for the request
-	isUser := ucred != nil
-	isSnap := (ucred != nil && ucred.Socket == dirs.SnapSocket)
-
-	// ensure that snaps can only access SnapOK things
-	if isSnap {
-		if c.SnapOK {
-			return accessOK
-		}
-		return accessUnauthorized
-	}
-
-	// the !RootOnly check is redundant, but belt-and-suspenders
-	if r.Method == "GET" && !c.RootOnly {
-		// Guest and user access restricted to GET requests
-		if c.GuestOK {
-			return accessOK
-		}
-
-		if isUser && c.UserOK {
-			return accessOK
-		}
-	}
-
-	// Remaining admin checks rely on identifying peer uid
-	if !isUser {
-		return accessUnauthorized
-	}
-
-	if ucred.Uid == 0 {
-		// Superuser does anything.
-		return accessOK
-	}
-
-	if c.RootOnly {
-		return accessUnauthorized
-	}
-
-	if c.PolkitOK != "" {
-		return checkPolkitAction(r, ucred, c.PolkitOK)
-	}
+	// Access control.
+	ReadAccess  accessChecker
+	WriteAccess accessChecker
```

Comment:
An idea for a test. I think it would be useful to have a test similar to TestAddRoutes that walks all registered API commands and checks that WriteAccess is not nil iff PUT || POST are set, perhaps even verify that it is not an instance of openAccess.

Reply:
We've essentially got that check here in `daemon_test.go`:

https://github.com/snapcore/snapd/blob/10f22a2a230e711042c61bd8d536e29fb56ed330/daemon/daemon_test.go#L343-L350

It doesn't specifically check for WriteAccess being set to openAccess, but it ensures that any command with a GET method sets ReadAccess, and any command with PUT or POST methods sets WriteAccess.